catalog/catalog_products/views.py

```diff
@@ -0,0 +1,105 @@
+from typing import Dict
+from django.http.response import JsonResponse
+from rest_framework.generics import CreateAPIView
+from rest_framework.views import APIView
+from catalog_products.serializers import CategorySerializer, ProductSerializer
+from catalog_products.models import Category, Product
+from rest_framework import viewsets
+
+
+class CategoryAllViewSet(viewsets.ModelViewSet):
+    serializer_class = CategorySerializer
+    queryset = Category.objects.all()
+
+
+class ProductAllViewSet(viewsets.ModelViewSet):
+    serializer_class = ProductSerializer
+    queryset = Product.objects.all()
+
+
+class Task1ViewSet(APIView):
+    queryset = Category.objects.all()
+    serializer_class = CategorySerializer
+    def get(self, request):
+        category_all = Category.objects.all()
+        category_without_parent = []
+        category_with_parent = []
+        for category in category_all:
+            if category.parent != None:
+                category_with_parent.append(category)
+            else:
+                category_without_parent.append(category)
+        category_name_with_parent = list(map(lambda category: f'Subcategory: {category.name}, parent category: {category.parent.name}',category_with_parent))
+        category_name_without_parent = list(map(lambda category: f'Subcategory: {category.name} havent parent category!',category_without_parent))      
+        return JsonResponse(category_name_with_parent + category_name_without_parent, safe=False)
+
+
+class Task2ViewSet(APIView):
+    queryset = Category.objects.all()
+    serializer_class = CategorySerializer
+
+    def get(self, request):
+        def get_count_recoursively(category):
+            category_all = Category.objects.all()
+            subcategory_count = 0
+            for cat in category_all:
+                if cat.parent == category:
+                    subcategory_count += 1
+                    subcategory_count += get_count_recoursively(cat)
+            return subcategory_count
+
+        category_all = Category.objects.all()
+        result = []
+        for category in category_all:
+            result.append(f'Category: {category.name} - Subcategory: {get_count_recoursively(category)}')
```

Comment:
If we used a database that supports recursive selects we would probably write a raw sql query in order to optimize our code and make only one request rather than more.

Take a look here:
https://dev.to/shawara/django-orm-optimization-tips-4-recursive-query-4on1